// e2e4/src/contracts/operation-status.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Represents possible values for operation status (e.g. of request to the server).
 */
var OperationStatus;
(function (OperationStatus) {
    /**
     * Nothing was performed before.
     */
    OperationStatus[OperationStatus["Initial"] = 0] = "Initial";
    /**
     * Last operation completed successfully.
     */
    OperationStatus[OperationStatus["Done"] = 1] = "Done";
    /**
     * Operation is performing right now.
     */
    OperationStatus[OperationStatus["Progress"] = 2] = "Progress";
    /**
     * Last operation completed with failure.
     */
    OperationStatus[OperationStatus["Fail"] = 3] = "Fail";
    /**
     * Last operation was cancelled.
     */
    OperationStatus[OperationStatus["Cancelled"] = 4] = "Cancelled";
    /**
     * Last operation doesn't return any data.
     */
    OperationStatus[OperationStatus["NoData"] = 5] = "NoData";
})(OperationStatus = exports.OperationStatus || (exports.OperationStatus = {}));
