// e2e4/esm/src/default-selection-service.js
/**
 * Default implementation of {@link SelectionService}.
 */
export class DefaultSelectionService {
    constructor() {
        /**
         * Collection of {@link SelectionTuple} elements which represents currently selected items in {@link items} collection.
         */
        this.selectionsList = new Array();
        /**
         * Default tracking function that will be used if nothing was specified for {@link trackByFn}.
         * Implements comparison by reference equality of objects.
         */
        this.trackByIdentity = (index, item) => item;
        this.trackByFn = this.trackByIdentity;
    }
    /**
     * @inheritdoc
     */
    destroy() {
        this.selectionsList.length = 0;
        this.lastProcessedIndex = null;
        this.items = null;
    }
    /**
     * @inheritdoc
     */
    checkSelection() {
        if (this.items !== null && typeof this.items !== 'undefined') {
            for (let i = this.selectionsList.length - 1; i >= 0; i--) {
                const tuple = this.selectionsList[i];
                const trackFn = this.trackByFn || this.trackByIdentity;
                if (this.isIndexAcceptable(tuple.index) &&
                    trackFn(tuple.index, this.items[tuple.index]) === trackFn(tuple.index, tuple.item)) {
                    tuple.item = this.items[tuple.index];
                    this.selectItem(tuple, true);
                }
                else {
                    this.deselectItem(tuple);
                }
            }
        }
        else {
            this.deselectAll();
        }
    }
    /**
     * Checks that applied index is valid number and it's value is inside {@link items} boundaries.
     * @param index index to check.
     * @returns `true` if index is valid.
     */
    isIndexAcceptable(index) {
        return index !== null && typeof index !== 'undefined' && index >= 0 && this.items && this.items.length > index;
    }
    /**
     * @inheritdoc
     */
    deselectAll() {
        const list = this.selectionsList.splice(0, this.selectionsList.length);
        for (const item of list) {
            this.processSelection(item, false);
        }
        this.lastProcessedIndex = null;
    }
    /**
     * @inheritdoc
     */
    selectAll() {
        this.selectRange(0, this.items.length - 1);
    }
    /**
     * @inheritdoc
     */
    selectRange(fromIndex, toIndex) {
        if (toIndex < 0 || this.items.length <= toIndex || fromIndex < 0 || this.items.length <= fromIndex) {
            return;
        }
        const startIndex = Math.min(fromIndex, toIndex);
        const endIndex = Math.max(fromIndex, toIndex);
        if (this.isRangeSelected(startIndex, endIndex)) {
            return;
        }
        this.deselectAll();
        const tempData = new Array();
        for (let i = startIndex; i <= endIndex; i++) {
            const tuple = this.getSelectionTuple(i);
            tempData.push(tuple);
            this.processSelection(tuple, true);
        }
        this.selectionsList.splice(0, this.selectionsList.length, ...tempData);
        this.lastProcessedIndex = endIndex;
    }
    /**
     * @inheritdoc
     */
    hasSelections() {
        return this.selectionsList.length !== 0;
    }
    /**
     * @inheritdoc
     */
    isRangeSelected(from, to) {
        // nothing selected
        if (this.selectionsList.length === 0) {
            return false;
        }
        // entire list selected
        if (from === 0 && to === this.items.length - 1 && this.selectionsList.length === this.items.length) {
            return true;
        }
        const orderedIndexes = this.selectionsList.map((tuple) => tuple.index).sort();
        return (1 + to - from === orderedIndexes.length &&
            orderedIndexes[0] === from &&
            orderedIndexes[orderedIndexes.length - 1] === to);
    }
    /**
     * @inheritdoc
     */
    isIndexSelected(index) {
        if (this.selectionsList.length > 0 && this.isIndexAcceptable(index)) {
            return this.selectionsList.findIndex((st) => st.index === index) !== -1;
        }
        return false;
    }
    /**
     * @inheritdoc
     */
    getItemIndex(item) {
        return this.items.findIndex((value) => value === item);
    }
    /**
     * @inheritdoc
     */
    getMinSelectedIndex() {
        let minIndex = -1;
        this.selectionsList.forEach((item) => {
            minIndex = minIndex === -1 || item.index < minIndex ? item.index : minIndex;
        });
        return minIndex;
    }
    /**
     * @inheritdoc
     */
    getMaxSelectedIndex() {
        let maxIndex = -1;
        this.selectionsList.forEach((item) => {
            maxIndex = maxIndex === -1 || item.index > maxIndex ? item.index : maxIndex;
        });
        return maxIndex;
    }
    /**
     * @inheritdoc
     */
    selectFirst() {
        if (this.items.length > 0) {
            this.selectItem(this.getSelectionTuple(0));
        }
    }
    /**
     * @inheritdoc
     */
    selectLast() {
        if (this.items.length > 0) {
            this.selectItem(this.getSelectionTuple(this.items.length - 1));
        }
    }
    /**
     * @inheritdoc
     */
    selectIndex(index, savePrevious = false) {
        if (this.isIndexAcceptable(index)) {
            this.selectItem(this.getSelectionTuple(index), savePrevious);
        }
    }
    /**
     * @inheritdoc
     */
    deselectIndex(index) {
        if (this.isIndexAcceptable(index)) {
            this.deselectItem(this.getSelectionTuple(index));
        }
    }
    /**
     * @inheritdoc
     */
    toggleSelection(index, savePrevious = false) {
        if (!this.isIndexAcceptable(index)) {
            return;
        }
        const tuple = this.getSelectionTuple(index);
        if (this.isIndexSelected(index) &&
            (this.selectionsList.length === 1 || (this.selectionsList.length > 1 && savePrevious))) {
            this.deselectItem(tuple);
            return;
        }
        this.selectItem(tuple, savePrevious);
    }
    /**
     * @inheritdoc
     */
    getSelectedElements() {
        return this.selectionsList.map((selectionTuple) => selectionTuple.item);
    }
    /**
     * @inheritdoc
     */
    getSelectedIndexes() {
        return this.selectionsList.map((selectionTuple) => selectionTuple.index);
    }
    /**
     * Performs final processing of selection/deselection of element.
     *
     * Current implementation sets {@link selected} propery of element (if it's defined).
     */
    processSelection(tuple, selected) {
        if (Object.prototype.hasOwnProperty.call(tuple.item, 'selected')) {
            tuple.item.selected = selected;
        }
    }
    /**
     * Internal method that used to perform item selection.
     */
    deselectItem(selectionTuple) {
        const index = this.selectionsList.findIndex((selectedItem) => selectedItem.item === selectionTuple.item);
        if (index !== -1) {
            this.selectionsList.splice(index, 1);
        }
        this.processSelection(selectionTuple, false);
        this.lastProcessedIndex = selectionTuple.index;
    }
    /**
     * Internal method that used to execute item deselection.
     */
    selectItem(selectionTuple, savePrevious = false) {
        if (savePrevious) {
            const index = this.selectionsList.findIndex((selectedItem) => selectedItem.item === selectionTuple.item);
            if (index !== -1) {
                this.selectionsList.splice(index, 1);
            }
            this.selectionsList.push(selectionTuple);
            this.processSelection(selectionTuple, true);
        }
        else {
            const list = this.selectionsList.splice(0, this.selectionsList.length);
            list.forEach((selectedItem) => {
                this.processSelection(selectedItem, false);
            });
            this.selectionsList.push(selectionTuple);
            this.processSelection(selectionTuple, true);
        }
        this.lastProcessedIndex = selectionTuple.index;
    }
    /**
     * Internal method that used to represent selection item as {@link SelectionTuple}.
     */
    getSelectionTuple(index) {
        return {
            index,
            item: this.items[index]
        };
    }
}
