// e2e4/esm/src/list.js
import { OperationStatus } from './contracts/operation-status';
import { NullObjectPager } from './null-object-pager';
import { destroyAll } from './utilities';
export class List {
    constructor(asyncSubscriber, stateServices, sortingsService, filtersService) {
        this.asyncSubscriber = asyncSubscriber;
        this.sortingsService = sortingsService;
        this.filtersService = filtersService;
        /**
         * Specifies that list must destroy previously loaded records immediately or keep them until data request is completed.
         */
        this.keepRecordsOnLoad = List.settings.keepRecordsOnLoad;
        /**
         * Array of elements transferred in the {@link ListResponse.items} property.
         */
        this.items = new Array();
        /**
         * Array of registered {@link StateService} instances.
         */
        this.stateServices = new Array();
        this.statusInternal = OperationStatus.Initial;
        this.destroyedInternal = false;
        this.initedInternal = false;
        this.loadDataSuccessCallback = (response) => {
            if (this.tryInterceptStatusResponse(response)) {
                return response;
            }
            this.tryCleanItemsOnLoad(true);
            return this.loadSuccessCallback(response);
        };
        this.reloadDataSuccessCallback = (response) => {
            if (this.tryInterceptStatusResponse(response)) {
                return response;
            }
            this.tryCleanItemsOnReload(true);
            return this.loadSuccessCallback(response);
        };
        this.loadDataFailCallback = () => {
            this.tryCleanItemsOnLoad(true);
            this.loadFailCallback();
        };
        this.reloadDataFailCallback = () => {
            this.tryCleanItemsOnReload(true);
            this.loadFailCallback();
        };
        if (stateServices != null) {
            if (Array.isArray(stateServices)) {
                this.stateServices.push(...(stateServices));
            }
            else {
                this.stateServices.push(stateServices);
            }
        }
        this.pager = new NullObjectPager();
    }
    /**
     * Configured {@link Pager} service.
     */
    get pager() {
        return this.pagerInternal;
    }
    set pager(value) {
        this.filtersService.removeFilterTarget(this.pagerInternal);
        this.pagerInternal = value;
        this.filtersService.registerFilterTarget(this.pagerInternal);
    }
    /**
     * True if the service was already destroyed via {@link destroy} call.
     */
    get destroyed() {
        return this.destroyedInternal;
    }
    /**
     * True if the service was already initialized via {@link init} call.
     */
    get inited() {
        return this.initedInternal;
    }
    /**
     * Current execution status of the list.
     */
    get status() {
        return this.statusInternal;
    }
    /**
     * returns `true`, if there is a data request executed at the moment (i.e. {@link state} is equal to {@link ProgressState.Progress})
     */
    get busy() {
        return this.status === OperationStatus.Progress;
    }
    /**
     * returns `true`, if there is no data request executed at the moment (i.e. {@link state} is NOT equal to {@link ProgressState.Progress})
     */
    get ready() {
        return this.status !== OperationStatus.Progress;
    }
    /**
     * Performs initialization logic of the service. This method must be called before first use of the service.
     */
    init() {
        if (this.inited) {
            return;
        }
        this.filtersService.registerFilterTarget(this.pager, this.sortingsService);
        const restoredState = {};
        Object.assign(restoredState, ...this.stateServices.map((service) => service.getState() || {}));
        this.filtersService.applyParams(restoredState);
        this.initedInternal = true;
    }
    /**
     * Performs destroy logic of the list itself and all of the inner services.
     */
    destroy() {
        this.asyncSubscriber.destroy();
        this.filtersService.destroy();
        this.sortingsService.destroy();
        this.clearData();
        this.destroyedInternal = true;
    }
    /**
     * Registers passed object(s) as state service to manage the list state.
     */
    registerStateService(...services) {
        services.forEach((service) => {
            this.stateServices.push(service);
        });
    }
    /**
     * Removes passed object(s) from state services collection of the list.
     */
    removeStateService(...services) {
        services.forEach((service) => {
            const index = this.stateServices.findIndex((s) => s === service);
            if (index !== -1) {
                this.stateServices.splice(index, 1);
            }
        });
    }
    /**
     * Registers passed object(s) as filter targets in underlying {@link FiltersService} to include their configured properties as parameters to the data request.
     * @see {@link FiltersService.registerFilterTarget}
     */
    registerFilterTarget(...targets) {
        this.filtersService.registerFilterTarget(...targets);
    }
    /**
     * @see {@link FiltersService.removeFilterTarget}
     */
    removeFilterTarget(...targets) {
        this.filtersService.removeFilterTarget(...targets);
    }
    /**
     * @see {@link FiltersService.getRequestState}
     */
    getRequestState(filterFn) {
        return this.filtersService.getRequestState(filterFn);
    }
    /**
     * Resets the list parameters (sortings, paging, filters) to their default values.
     */
    resetSettings() {
        this.filtersService.resetValues();
    }
    /**
     * Cancels the request executed at the moment.
     */
    cancelRequests() {
        if (this.busy) {
            this.asyncSubscriber.detach();
            this.statusInternal = OperationStatus.Cancelled;
            this.clearData();
        }
    }
    /**
     * Clears {@link items} array. Calls {@link destroyAll} method for {@link items} array to perform optional destroy logic of the elements.
     * {@see destroyAll}
     */
    clearData() {
        destroyAll(this.items);
        this.items = [];
    }
    /**
     * Performs data loading by calling specified {@link fetchMethod} delegate.
     * @return result of {@link fetchMethod} execution.
     */
    loadData() {
        if (this.busy) {
            return null;
        }
        const subscribable = this.beginRequest();
        this.tryCleanItemsOnLoad(false);
        this.asyncSubscriber.attach(subscribable, this.loadDataSuccessCallback, this.loadDataFailCallback);
        this.stateServices.forEach((service) => {
            service.persistState(this.filtersService);
        });
        return subscribable;
    }
    /**
     * Resets paging parameters and performs data loading by calling {@link loadData} if list not in {@link OperationStatus.Progress} state.
     * @return result of {@link fetchMethod} if it was called. `null` otherwise.
     */
    reloadData() {
        if (this.busy) {
            return null;
        }
        this.pager.reset();
        const subscribable = this.beginRequest();
        this.tryCleanItemsOnReload(false);
        this.asyncSubscriber.attach(subscribable, this.reloadDataSuccessCallback, this.reloadDataFailCallback);
        this.stateServices.forEach((service) => {
            service.persistState(this.filtersService);
        });
        return subscribable;
    }
    /**
     * Callback which is executed if {@link fetchMethod} execution finished successfully.
     */
    loadSuccessCallback(response) {
        const items = Array.isArray(response) ? response : response.items;
        this.items = this.items.concat(items);
        this.pager.processResponse(response);
        // In case when filter changed from last request and there's no data now
        // Don't do this for flat responses since we don't know real count items
        if (!Array.isArray(response) && this.pager.totalCount === 0) {
            this.clearData();
            this.pager.reset();
        }
        this.statusInternal =
            this.pager.totalCount === 0 && this.items.length === 0 ? OperationStatus.NoData : OperationStatus.Done;
        return response;
    }
    /**
     * Callback which is executed if {@link fetchMethod} execution finished with error.
     */
    loadFailCallback() {
        this.tryCleanItemsOnLoad(true);
        this.statusInternal = OperationStatus.Fail;
    }
    tryInterceptStatusResponse(response) {
        if (this.responseHasStatus(response)) {
            switch (response.status) {
                case OperationStatus.Fail:
                    this.reloadDataFailCallback();
                    return true;
                case OperationStatus.Cancelled:
                    this.cancelRequests();
                    return true;
                case OperationStatus.Progress:
                    return true;
            }
        }
        return false;
    }
    tryCleanItemsOnLoad(requestCompleted) {
        if (this.keepRecordsOnLoad === requestCompleted && this.pager.appendedOnLoad === false) {
            this.clearData();
        }
    }
    tryCleanItemsOnReload(requestCompleted) {
        if (this.keepRecordsOnLoad === requestCompleted) {
            this.clearData();
        }
    }
    beginRequest() {
        this.statusInternal = OperationStatus.Progress;
        const requestState = this.filtersService.getRequestState();
        return this.fetchMethod(requestState);
    }
    responseHasStatus(response) {
        return (response.status !== null &&
            typeof response.status !== 'undefined');
    }
}
/**
 * Global settings of list..
 *
 * These settings are static and their values are copied to the properties of the same name for each instance of {@link List} type.
 *
 * So, changing of this settings will affect all instances of {@link List} type that will be created after such changes.
 * If you want to change settings of concrete object you can use it the same name properties.
 */
// tslint:disable-next-line: typedef
List.settings = {
    /**
     * @see {@link List.keepRecordsOnLoad}
     */
    keepRecordsOnLoad: false
};
