// e2e4/esm/src/null-object-pager.js
/**
 * Implements {@link Pager} contract and represents list without any paging mechanics.
 * @note This type is configured to use with {@link FiltersService}.
 */
export class NullObjectPager {
    constructor() {
        /**
         * @inheritdoc
         */
        this.appendedOnLoad = false;
        /**
         * @inheritdoc
         */
        this.totalCount = 0;
        /**
         * @inheritdoc
         */
        this.loadedCount = 0;
    }
    /**
     * @inheritdoc
     */
    processResponse(response) {
        let alignedResponse;
        if (Array.isArray(response)) {
            alignedResponse = {
                items: response,
                loadedCount: response.length,
                totalCount: response.length
            };
        }
        else {
            alignedResponse = response;
        }
        this.loadedCount =
            alignedResponse.loadedCount ||
                (alignedResponse.items && alignedResponse.items.length ? alignedResponse.items.length : 0);
        this.totalCount = alignedResponse.totalCount || 0;
    }
    /**
     * @inheritdoc
     */
    reset() {
        this.totalCount = 0;
        this.loadedCount = 0;
    }
}
