// e2e4/esm/src/sortings-service.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var __metadata = (this && this.__metadata) || function (k, v) {
    if (typeof Reflect === "object" && typeof Reflect.metadata === "function") return Reflect.metadata(k, v);
};
import { SortDirection } from './contracts/sort-parameter';
import { filter } from './filter-annotation';
/**
 * Provides sorting functionality.
 * @note This type is configured to use with {@link FiltersService}.
 */
export class SortingsService {
    constructor() {
        /**
         * Sortings that were selected by the user and must be applied on next request of data.
         *
         * @note This property is ready to use with {@link FiltersService} since it has {@link filter} annotation.
         */
        this.sortings = new Array();
        /**
         * Internal implementation of {@link defaultSortings}.
         */
        this.defaultSortingsInternal = new Array();
    }
    /**
     * Default sortings that will be used by service.
     */
    get defaultSortings() {
        return this.defaultSortingsInternal;
    }
    /**
     * If called when {@link sortings} is empty then applied value will be copied to {@link sortings} immediately.
     */
    set defaultSortings(value) {
        this.defaultSortingsInternal = value || [];
        if (this.sortings.length === 0) {
            this.sortings = this.cloneDefaultSortings();
        }
    }
    /**
     * Sets {@link sortings} according to specified parameters.
     * @param fieldName name of the field by which sorting must be executed on server. This value will be used as {@link SortParameter.fieldName}.
     *
     * In case when sorting with the same field name is already specified, direction of this sorting will be toggled to reversed value and this sorting will be pushed to the end of {@link sortings} array.
     * So it will be applied last.
     * @param savePrevious `true` to keep previously applied sortings in {@link sortings} array.
     */
    setSort(fieldName, savePrevious) {
        let newSort = { direction: SortDirection.Asc, fieldName };
        for (let i = 0; i < this.sortings.length; i++) {
            if (this.sortings[i].fieldName === fieldName) {
                const existedSort = this.sortings.splice(i, 1)[0];
                newSort = {
                    direction: existedSort.direction,
                    fieldName: existedSort.fieldName
                };
                newSort.direction = newSort.direction === SortDirection.Asc ? SortDirection.Desc : SortDirection.Asc;
                break;
            }
        }
        if (savePrevious) {
            this.sortings.push(newSort);
        }
        else {
            this.sortings.length = 0;
            this.sortings.push(newSort);
        }
    }
    /**
     * Removes sort with specified field name  from {@link sortings} array.
     * @param fieldName name of the sort to remove.
     */
    removeSort(fieldName) {
        for (let i = 0; i < this.sortings.length; i++) {
            if (this.sortings[i].fieldName === fieldName) {
                this.sortings.splice(i, 1);
            }
        }
    }
    /**
     * Removes all sortings from {@link sortings} array.
     */
    removeAllSortings() {
        this.sortings.length = 0;
    }
    /**
     * Performs service destroy.
     */
    destroy() {
        this.defaultSortingsInternal.length = 0;
        this.sortings.length = 0;
    }
    /**
     * Internal method for default sortings cloning.
     * This method is used as {@link FilterConfig.defaultValue} as well as for copying to {@link sortings} when {@link defaultSortings} setter is used and {@link sortings} is empty.
     */
    cloneDefaultSortings() {
        return this.defaultSortingsInternal.map((s) => ({
            direction: s.direction,
            fieldName: s.fieldName
        }));
    }
}
__decorate([
    filter({
        defaultValue() {
            return this.cloneDefaultSortings();
        },
        parameterName: 'sortings',
        parseFormatter(rawValue) {
            return Array.isArray(rawValue)
                ? rawValue.map((sort) => ({
                    direction: sort.direction * 1,
                    fieldName: sort.fieldName
                }))
                : [];
        },
        serializeFormatter() {
            return this.sortings.map((sort) => ({
                direction: sort.direction,
                fieldName: sort.fieldName
            }));
        }
    }),
    __metadata("design:type", Array)
], SortingsService.prototype, "sortings", void 0);
