// e2e4/src/default-selection-service.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Default implementation of {@link SelectionService}.
 */
var DefaultSelectionService = /** @class */ (function () {
    function DefaultSelectionService() {
        /**
         * Collection of {@link SelectionTuple} elements which represents currently selected items in {@link items} collection.
         */
        this.selectionsList = new Array();
        /**
         * Default tracking function that will be used if nothing was specified for {@link trackByFn}.
         * Implements comparison by reference equality of objects.
         */
        this.trackByIdentity = function (index, item) { return item; };
        this.trackByFn = this.trackByIdentity;
    }
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.destroy = function () {
        this.selectionsList.length = 0;
        this.lastProcessedIndex = null;
        this.items = null;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.checkSelection = function () {
        if (this.items !== null && typeof this.items !== 'undefined') {
            for (var i = this.selectionsList.length - 1; i >= 0; i--) {
                var tuple = this.selectionsList[i];
                var trackFn = this.trackByFn || this.trackByIdentity;
                if (this.isIndexAcceptable(tuple.index) &&
                    trackFn(tuple.index, this.items[tuple.index]) === trackFn(tuple.index, tuple.item)) {
                    tuple.item = this.items[tuple.index];
                    this.selectItem(tuple, true);
                }
                else {
                    this.deselectItem(tuple);
                }
            }
        }
        else {
            this.deselectAll();
        }
    };
    /**
     * Checks that applied index is valid number and it's value is inside {@link items} boundaries.
     * @param index index to check.
     * @returns `true` if index is valid.
     */
    DefaultSelectionService.prototype.isIndexAcceptable = function (index) {
        return index !== null && typeof index !== 'undefined' && index >= 0 && this.items && this.items.length > index;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.deselectAll = function () {
        var list = this.selectionsList.splice(0, this.selectionsList.length);
        for (var _i = 0, list_1 = list; _i < list_1.length; _i++) {
            var item = list_1[_i];
            this.processSelection(item, false);
        }
        this.lastProcessedIndex = null;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.selectAll = function () {
        this.selectRange(0, this.items.length - 1);
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.selectRange = function (fromIndex, toIndex) {
        var _a;
        if (toIndex < 0 || this.items.length <= toIndex || fromIndex < 0 || this.items.length <= fromIndex) {
            return;
        }
        var startIndex = Math.min(fromIndex, toIndex);
        var endIndex = Math.max(fromIndex, toIndex);
        if (this.isRangeSelected(startIndex, endIndex)) {
            return;
        }
        this.deselectAll();
        var tempData = new Array();
        for (var i = startIndex; i <= endIndex; i++) {
            var tuple = this.getSelectionTuple(i);
            tempData.push(tuple);
            this.processSelection(tuple, true);
        }
        (_a = this.selectionsList).splice.apply(_a, [0, this.selectionsList.length].concat(tempData));
        this.lastProcessedIndex = endIndex;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.hasSelections = function () {
        return this.selectionsList.length !== 0;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.isRangeSelected = function (from, to) {
        // nothing selected
        if (this.selectionsList.length === 0) {
            return false;
        }
        // entire list selected
        if (from === 0 && to === this.items.length - 1 && this.selectionsList.length === this.items.length) {
            return true;
        }
        var orderedIndexes = this.selectionsList.map(function (tuple) { return tuple.index; }).sort();
        return (1 + to - from === orderedIndexes.length &&
            orderedIndexes[0] === from &&
            orderedIndexes[orderedIndexes.length - 1] === to);
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.isIndexSelected = function (index) {
        if (this.selectionsList.length > 0 && this.isIndexAcceptable(index)) {
            return this.selectionsList.findIndex(function (st) { return st.index === index; }) !== -1;
        }
        return false;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.getItemIndex = function (item) {
        return this.items.findIndex(function (value) { return value === item; });
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.getMinSelectedIndex = function () {
        var minIndex = -1;
        this.selectionsList.forEach(function (item) {
            minIndex = minIndex === -1 || item.index < minIndex ? item.index : minIndex;
        });
        return minIndex;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.getMaxSelectedIndex = function () {
        var maxIndex = -1;
        this.selectionsList.forEach(function (item) {
            maxIndex = maxIndex === -1 || item.index > maxIndex ? item.index : maxIndex;
        });
        return maxIndex;
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.selectFirst = function () {
        if (this.items.length > 0) {
            this.selectItem(this.getSelectionTuple(0));
        }
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.selectLast = function () {
        if (this.items.length > 0) {
            this.selectItem(this.getSelectionTuple(this.items.length - 1));
        }
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.selectIndex = function (index, savePrevious) {
        if (savePrevious === void 0) { savePrevious = false; }
        if (this.isIndexAcceptable(index)) {
            this.selectItem(this.getSelectionTuple(index), savePrevious);
        }
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.deselectIndex = function (index) {
        if (this.isIndexAcceptable(index)) {
            this.deselectItem(this.getSelectionTuple(index));
        }
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.toggleSelection = function (index, savePrevious) {
        if (savePrevious === void 0) { savePrevious = false; }
        if (!this.isIndexAcceptable(index)) {
            return;
        }
        var tuple = this.getSelectionTuple(index);
        if (this.isIndexSelected(index) &&
            (this.selectionsList.length === 1 || (this.selectionsList.length > 1 && savePrevious))) {
            this.deselectItem(tuple);
            return;
        }
        this.selectItem(tuple, savePrevious);
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.getSelectedElements = function () {
        return this.selectionsList.map(function (selectionTuple) { return selectionTuple.item; });
    };
    /**
     * @inheritdoc
     */
    DefaultSelectionService.prototype.getSelectedIndexes = function () {
        return this.selectionsList.map(function (selectionTuple) { return selectionTuple.index; });
    };
    /**
     * Performs final processing of selection/deselection of element.
     *
     * Current implementation sets {@link selected} propery of element (if it's defined).
     */
    DefaultSelectionService.prototype.processSelection = function (tuple, selected) {
        if (Object.prototype.hasOwnProperty.call(tuple.item, 'selected')) {
            tuple.item.selected = selected;
        }
    };
    /**
     * Internal method that used to perform item selection.
     */
    DefaultSelectionService.prototype.deselectItem = function (selectionTuple) {
        var index = this.selectionsList.findIndex(function (selectedItem) { return selectedItem.item === selectionTuple.item; });
        if (index !== -1) {
            this.selectionsList.splice(index, 1);
        }
        this.processSelection(selectionTuple, false);
        this.lastProcessedIndex = selectionTuple.index;
    };
    /**
     * Internal method that used to execute item deselection.
     */
    DefaultSelectionService.prototype.selectItem = function (selectionTuple, savePrevious) {
        var _this = this;
        if (savePrevious === void 0) { savePrevious = false; }
        if (savePrevious) {
            var index = this.selectionsList.findIndex(function (selectedItem) { return selectedItem.item === selectionTuple.item; });
            if (index !== -1) {
                this.selectionsList.splice(index, 1);
            }
            this.selectionsList.push(selectionTuple);
            this.processSelection(selectionTuple, true);
        }
        else {
            var list = this.selectionsList.splice(0, this.selectionsList.length);
            list.forEach(function (selectedItem) {
                _this.processSelection(selectedItem, false);
            });
            this.selectionsList.push(selectionTuple);
            this.processSelection(selectionTuple, true);
        }
        this.lastProcessedIndex = selectionTuple.index;
    };
    /**
     * Internal method that used to represent selection item as {@link SelectionTuple}.
     */
    DefaultSelectionService.prototype.getSelectionTuple = function (index) {
        return {
            index: index,
            item: this.items[index]
        };
    };
    return DefaultSelectionService;
}());
exports.DefaultSelectionService = DefaultSelectionService;
