// e2e4/src/null-object-pager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Implements {@link Pager} contract and represents list without any paging mechanics.
 * @note This type is configured to use with {@link FiltersService}.
 */
var NullObjectPager = /** @class */ (function () {
    function NullObjectPager() {
        /**
         * @inheritdoc
         */
        this.appendedOnLoad = false;
        /**
         * @inheritdoc
         */
        this.totalCount = 0;
        /**
         * @inheritdoc
         */
        this.loadedCount = 0;
    }
    /**
     * @inheritdoc
     */
    NullObjectPager.prototype.processResponse = function (response) {
        var alignedResponse;
        if (Array.isArray(response)) {
            alignedResponse = {
                items: response,
                loadedCount: response.length,
                totalCount: response.length
            };
        }
        else {
            alignedResponse = response;
        }
        this.loadedCount =
            alignedResponse.loadedCount ||
                (alignedResponse.items && alignedResponse.items.length ? alignedResponse.items.length : 0);
        this.totalCount = alignedResponse.totalCount || 0;
    };
    /**
     * @inheritdoc
     */
    NullObjectPager.prototype.reset = function () {
        this.totalCount = 0;
        this.loadedCount = 0;
    };
    return NullObjectPager;
}());
exports.NullObjectPager = NullObjectPager;
