// e2e4/src/operation.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var operation_status_1 = require("./contracts/operation-status");
/**
 * Represents status of operation that must be displayed on UI inside {@link StatusTrackingService}.
 */
var Operation = /** @class */ (function () {
    /**
     * @param status value for {@link status} property.
     * @param title value for {@link title} property.
     */
    function Operation(status, title) {
        this.status = status;
        this.title = title;
    }
    Object.defineProperty(Operation.prototype, "className", {
        /**
         * Returns class name from {@link Operation.settings} that corresponds to current {@link status} value.
         */
        get: function () {
            switch (this.status) {
                case operation_status_1.OperationStatus.Done:
                    return Operation.settings.statusDoneClassName;
                case operation_status_1.OperationStatus.Progress:
                    return Operation.settings.statusProgressClassName;
                case operation_status_1.OperationStatus.Fail:
                    return Operation.settings.statusFailClassName;
                default:
                    return '';
            }
        },
        enumerable: true,
        configurable: true
    });
    /**
     * Settings for names of css classes that are relevant to {@link status} value.
     * This settings is used by {@link className} property and changing of this settings affects every object of {@link Operation} type.
     */
    // tslint:disable-next-line: typedef
    Operation.settings = {
        /**
         * Name of css class that corresponds to {@link OperationStatus.Done}.
         */
        statusDoneClassName: 'status-done',
        /**
         * Name of css class that corresponds to {@link OperationStatus.Fail}.
         */
        statusFailClassName: 'status-fail',
        /**
         * Name of css class that corresponds to {@link OperationStatus.Progress}.
         */
        statusProgressClassName: 'status-progress'
    };
    return Operation;
}());
exports.Operation = Operation;
